RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterECLCC.h

```diff
@@ -0,0 +1,297 @@
+        const int end = pfClusteringEdgeVars[v + 1].pfrh_edgeIdx();
+        int m = v;
+        int i = beg;
+        while ((m == v) && (i < end)) {
+          m = std::min(m, pfClusteringEdgeVars[i].pfrh_edgeList());
+          i++;
+        }
+        pfClusteringVars[v].pfrh_topoId() = m;
+      }
+
+      if (from == 0) {
+        pfClusteringVars.topL() = 0;
+        pfClusteringVars.posL() = 0;
+        pfClusteringVars.topH() = nRH - 1;
+        pfClusteringVars.posH() = nRH - 1;
+      }
+    }
+  };
+
+  // First edge processing kernel of ECL-CC
+  class ECLCCCompute1 {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFClusteringEdgeVarsDeviceCollection::View pfClusteringEdgeVars) const {
+      const int nRH = pfRecHits.size();
+
+      for (int v : cms::alpakatools::elements_with_stride(acc, nRH)) {
+        const int vstat = pfClusteringVars[v].pfrh_topoId();
+        if (v != vstat) {
+          const int beg = pfClusteringEdgeVars[v].pfrh_edgeIdx();
+          const int end = pfClusteringEdgeVars[v + 1].pfrh_edgeIdx();
+          int deg = end - beg;
+          if (deg > 16) {
+            int idx;
+            if (deg <= 352) {
+              idx = alpaka::atomicAdd(acc, &pfClusteringVars.topL(), 1);
+            } else {
+              idx = alpaka::atomicAdd(acc, &pfClusteringVars.topH(), -1);
+            }
+            pfClusteringVars[idx].wl_d() = v;
+          } else {
+            int vstat = representative(v, pfClusteringVars);
+            for (int i = beg; i < end; i++) {
+              const int nli = pfClusteringEdgeVars[i].pfrh_edgeList();
+              if (v > nli) {
+                int ostat = representative(nli, pfClusteringVars);
+                bool repeat;
+                do {
+                  repeat = false;
+                  if (vstat != ostat) {
+                    int ret;
+                    if (vstat < ostat) {
+                      if ((ret = alpaka::atomicCas(acc, &pfClusteringVars[ostat].pfrh_topoId(), ostat, vstat)) !=
+                          ostat) {
+                        ostat = ret;
+                        repeat = true;
+                      }
+                    } else {
+                      if ((ret = alpaka::atomicCas(acc, &pfClusteringVars[vstat].pfrh_topoId(), vstat, ostat)) !=
+                          vstat) {
+                        vstat = ret;
+                        repeat = true;
+                      }
+                    }
+                  }
+                } while (repeat);
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  /* process medium-degree vertices at warp granularity */
```

Comment:
how does this behave on cpu ?

Reply:
I may just remove `Compute2` and `Compute3` since I see no real detriment to throughput by not running them with Alpaka now. The nature of the inputs in our case is such that our vertex degree is always 8 (8 neighbors) so these will never be needed.